ai/vllm-deployment/hpa/README.md

```diff
@@ -0,0 +1,64 @@
+# Horizontal Pod Autoscaling AI Inference Server
+
+This exercise shows how to set up the infrastructure to automatically
+scale an AI inference server, using custom metrics (either server
+or GPU metrics). This exercise requires a running Prometheus instance,
+preferably managed by the Prometheus Operator. We assume
+you already have the vLLM AI inference server running from this
+[exercise](../README.md), in the parent directory.
+
+## Architecture
+
+The autoscaling solution works as follows:
+
+1.  The **vLLM Server** or the **NVIDIA DCGM Exporter** exposes raw metrics on a `/metrics` endpoint.
+2.  A **ServiceMonitor** resource declaratively specifies how Prometheus should discover and scrape these metrics.
+3.  The **Prometheus Operator** detects the `ServiceMonitor` and configures its managed **Prometheus Server** instance to begin scraping the metrics.
+4.  For GPU metrics, a **PrometheusRule** is used to relabel the raw DCGM metrics, creating a new, HPA-compatible metric.
+5.  The **Prometheus Adapter** queries the Prometheus Server for the processed metrics and exposes them through the Kubernetes custom metrics API.
+6.  The **Horizontal Pod Autoscaler (HPA)** controller queries the custom metrics API for the metrics and compares them to the target values defined in the `HorizontalPodAutoscaler` resource.
+7.  If the metrics exceed the target, the HPA scales up the `vllm-gemma-deployment`.
+
+```
+┌──────────────┐   ┌────────────────┐   ┌──────────────────┐
+│ User Request │──>│ vLLM Server    │──>│ ServiceMonitor   │
+└──────────────┘   │ (or DCGM Exp.) │   └──────────────────┘
+                   └────────────────┘            │
+                                                 ▼
+┌────────────────┐   ┌──────────────────┐   ┌──────────────────┐
+│ HPA Controller │<──│ Prometheus Adpt. │<──│ Prometheus Srv.  │
+└────────────────┘   └──────────────────┘   └──────────────────┘
+                                                 │ (GPU Path Only)
+                                                 ▼
+                                           ┌────────────────┐
+                                           │ PrometheusRule │
+                                           └────────────────┘
```

Comment:
Some suggestions for the diagram to make it more clear:
* Use numbered steps and arrow directions to guide the user through the precise data flow (scrape, evaluate, record, query, scale) from start to finish
* The flow hides the crucial transformation step where a raw metric is converted into a processed metric. Recommend to clearly label the initial scrape with the raw DCGM metric name and the query from the adapter with the new, processed metric name.
* The PrometheusRule is shown as a final step in the "GPU Path Only". However, the PrometheusRule is not a destination for data; it's a configuration that tells the Prometheus Server how to perform an internal calculation
* Include the Kubernetes API Server between the adapter and HPA

Reply:
Just noticed that GitHub supports mermaid diagrams in markdown: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams#creating-mermaid-diagrams
Might be easier to edit than ASCII diagrams.